src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Utilities/SpecializedTasks.cs

```diff
@@ -86,6 +87,60 @@ public static ValueTask<T[]> WhenAll<T>(IEnumerable<ValueTask<T>> tasks)
             }
         }
 
+        /// <summary>
+        /// This helper method provides semantics equivalent to the following, but avoids throwing an intermediate
+        /// <see cref="OperationCanceledException"/> in the case where the asynchronous operation is cancelled.
+        ///
+        /// <code><![CDATA[
+        /// public ValueTask<TResult> MethodAsync(TArg arg, CancellationToken cancellationToken)
+        /// {
+        ///   var intermediate = await func(arg, cancellationToken).ConfigureAwait(false);
+        ///   return transform(intermediate);
+        /// }
+        /// ]]></code>
+        /// </summary>
+        /// <remarks>
+        /// This helper method is only intended for use in cases where profiling reveals substantial overhead related to
+        /// cancellation processing.
+        /// </remarks>
+        /// <typeparam name="TArg">The type of a state variable to pass to <paramref name="func"/> and <paramref name="transform"/>.</typeparam>
+        /// <typeparam name="TIntermediate">The type of intermediate result produced by <paramref name="func"/>.</typeparam>
+        /// <typeparam name="TResult">The type of result produced by <paramref name="transform"/>.</typeparam>
+        /// <param name="func">The intermediate asynchronous operation.</param>
+        /// <param name="transform">The synchronous transformation to apply to the result of <paramref name="func"/>.</param>
+        /// <param name="arg">The state to pass to <paramref name="func"/> and <paramref name="transform"/>.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> that the operation will observe.</param>
+        /// <returns></returns>
+        public static ValueTask<TResult> TransformWithoutIntermediateCancellationExceptionAsync<TArg, TIntermediate, TResult>(
+            Func<TArg, CancellationToken, ValueTask<TIntermediate>> func,
+            Func<TIntermediate, TArg, TResult> transform,
+            TArg arg,
+            CancellationToken cancellationToken)
+        {
+            var intermediateResult = func(arg, cancellationToken);
+            if (intermediateResult.IsCompletedSuccessfully)
+            {
+                return new ValueTask<TResult>(transform(intermediateResult.Result, arg));
+            }
+            else if (intermediateResult.IsCanceled && cancellationToken.IsCancellationRequested)
+            {
+                return new ValueTask<TResult>(Task.FromCanceled<TResult>(cancellationToken));
+            }
+            else
+            {
+                return UnwrapAndTransformAsync(intermediateResult, transform, arg, cancellationToken);
+            }
+
+            static ValueTask<TResult> UnwrapAndTransformAsync(ValueTask<TIntermediate> intermediateResult, Func<TIntermediate, TArg, TResult> transform, TArg arg, CancellationToken cancellationToken)
```

Comment:
Do we need this local function?  Why not just inline it?

Also, can you doc what's going on here?

Reply:
Avoids capturing values for delegates on the fast path.

Reply:
Probably avoiding https://github.com/dotnet/roslyn/issues/20777

Reply:
thanks!